File: eftl/samples/python/shared_durable.py
'''
 Copyright (c) 2013-2020 TIBCO Software Inc.
 All Rights Reserved.
 For more information, please contact:
  TIBCO Software Inc., Palo Alto, California, USA
'''

import asyncio
import getopt, sys, time
import datetime
import threading
from messaging.eftl.connection import Eftl

user       = "user"
password   = "password"
url        = "ws://localhost:8585/channel"
connection = None
loop       = None
eFTL       = Eftl()
error      = False
clientId   = None
received   = 0
durName    = "sample-shared-durable"
ackMode    = "auto"
timer      = None
subscriberId = None

def usage():
    print (usage_lines)
    sys.exit(0)

usage_lines = """
        Usage: python3 shared_durable.py [options] url
        Options:
        \t-user        <str>    Username for authentication
        \t-password    <str>    Password for authentication
        \t-clientId    <str>    unique client id
        \t-name        <str>    durable name
        \t-count       <int>    Number of message to receive
      """

def parseArgs(argv):
    count = len(argv)
    i = 1

    while i < count:
        if (argv[i].lower() == '-user'):
            global user
            user = argv[i+1]
        elif (argv[i].lower() == "-password"):
            global password
            password = argv[i+1]
        elif (argv[i].lower() == "-count"):
            global recvCount
            recvCount = int(argv[i+1])
        elif (argv[i].lower() == "-clientid"):
            global clientId
            clientId = argv[i+1]
        elif (argv[i].lower() == "-name"):
            global durName
            durName = argv[i+1]
        elif (argv[i].lower() == "-help") or (argv[i].lower() == "-h"):
            usage()
        else:
           global url
           url = argv[i]

        i += 1

    print("shared_durable.py: " + eFTL.get_version())


# internal disconnect function
async def _disconnect():
    global connection

    if connection.is_connected():
        await connection.disconnect()

async def _unsubscribe(id):
    global connection

    await connection.unsubscribe(id)

def on_recv_timeout():
    global timer
    global subscriberId
    event_loop = None

    event_loop = asyncio.new_event_loop()

    timer.cancel()
    event_loop.run_until_complete(_unsubscribe(subscriberId))
    event_loop.run_until_complete(_disconnect())

async def on_subscribe(id):
    global timer
    global subscriberId 

    subscriberId = id
    print("Successfully subscribed to receive messages")

    timer = threading.Timer(30.0, on_recv_timeout)
    timer.start()

async def on_subscribe_error(code, reason):
    print("Failed to subscribe, code: %s, reason: %s" %(code, reason))
    
    if connection.is_connected():
        await _disconnect()

async def on_message(message):
    global received
    global recvCount
    print("Received message: [%s] %s" %(received, message))
    received += 1


# on_connect callback called on a successful connect
async def on_connect(connection):
    print("Connected to eFTL server: %s" % url)

    # change the matcher to "{}" if you wish to receive all messages
    matcher = "{\"type\":\"hello\"}"

    await connection.subscribe(matcher=matcher,
                              ack=ackMode,
                              durable=durName,
                              type="shared",
                              on_subscribe=on_subscribe,
                              on_error=on_subscribe_error,
                              on_message=on_message)

# on_reconnect callback
async def on_reconnect(connection):
    print("Reconnect attempt: ")

# on_error callback
async def on_error(connection, code, reason):
    global error
    print("Connection error [%s], reason %s" %(code, reason))
    error = True

# on_disconnect callback
async def on_disconnect(connection, loop, code, reason):
    print("Disconnected [%s]: %s" %(code, reason))

    loop.stop()

async def run():
    global connection
    global sendCount

    try:
        # connect to the eftl 
        connection = await eFTL.connect(url,
                                    user=user,
                                    password=password,
                                    trust_all=True,
                                    event_loop=loop,
                                    on_connect=on_connect,
                                    on_reconnect=on_reconnect,
                                    on_disconnect=on_disconnect,
                                    on_error=on_error)
    finally:
        print("Initiated connect..")
        

def cleanup():
    global loop
    global connection

    if connection.is_connected():
        loop.run_until_complete(_disconnect())

    loop.stop()
    loop.close()
    
def main(argv):
    errors = None
    global loop

    parseArgs(argv)

    try:
        loop = asyncio.get_event_loop();
        loop.run_until_complete(run())
        loop.run_forever()
    except (EOFError, KeyboardInterrupt):
        print("Disconnected from eFTL server: %s" % url)
    except Exception as err:
        errors = True
    finally:
           cleanup()
           if errors:
               sys.exit(1)
           else:
               sys.exit(0)

if __name__ == "__main__":
    main(sys.argv)

File: eftl/samples/dotnet/SharedDurable.cs
/*
 * Copyright (c) 2013-$Date: 2015-07-30 12:57:32 -0700 (Thu, 30 Jul 2015) $ Cloud Software Group, Inc.
 * All Rights Reserved.
 */

using System;
using System.Threading;
using System.Collections;

using TIBCO.EFTL;

public class Subscriber 
{
    static IConnection connection = null;

    static String url = "ws://localhost:8585/channel";
    static String username = "user";
    static String password = "password";
    static String durable = "sample-shared-durable";
    static String matcher = null;
    static int count = 10;
    static int received = 0;

    static CountdownEvent latch = new CountdownEvent(1);
    static Hashtable options = new Hashtable();

    private static void Usage() 
    {
        Console.WriteLine();
        Console.WriteLine("usage: Subscriber [options] url");
        Console.WriteLine();
        Console.WriteLine("options:");
        Console.WriteLine("  -u, --username <username>]");
        Console.WriteLine("  -p, --password <password>]");
        Console.WriteLine("  -n, --name <durable name>");
        Console.WriteLine("  -c, --count <count>");
        Console.WriteLine();
        System.Environment.Exit(1);
    }

    public class ConnectionListener: IConnectionListener
    {
        public void OnConnect(IConnection connection)
        {
            Console.WriteLine("Connected to eFTL server at: " + url);

            Subscriber.connection = connection;

            Subscriber.latch.Signal();
        }

        public void OnDisconnect(IConnection connection, int code, String reason)
        {
            Console.WriteLine("Disconnected from eFTL server: " + reason);

            Subscriber.latch.Signal();
        }

        public void OnError(IConnection connection, int code, String reason)
        {
            Console.WriteLine("Connection error: " + reason);

            Subscriber.latch.Signal();
        }

        public void OnReconnect(IConnection connection)
        {
            Console.WriteLine("Reconnected to eFTL server at: " + url);
        }
    } 

    public class SubscriptionListener: ISubscriptionListener
    {
        public void OnError(String subscriptionId, int code, String reason)
        {
            Console.WriteLine("Error subscribing to {0}: {1}", matcher, reason);
        }

        public void OnMessages(IMessage[] messages)
        {
            foreach (IMessage msg in messages)
            {
                received++;
                Console.WriteLine("Received message " + msg);
            }

            if (received >= count)
                Subscriber.latch.Signal();
        }

        public void OnSubscribe(String subscriptionId)
        {
            Console.WriteLine("Subscribed to {0}", (matcher != null ? matcher : "null matcher"));
        }
    }

    public Subscriber(String url, String username, String password) 
    {
        if (username != null) 
            options.Add(EFTL.PROPERTY_USERNAME, username);

        if (password != null) 
            options.Add(EFTL.PROPERTY_PASSWORD, password);
        
        // Asynchronously connect to the eFTL server.
        EFTL.Connect(url, options, new ConnectionListener());

        try {
            latch.Wait();
        } catch (Exception) {}
    }

    public bool IsConnected()
    {
       return Subscriber.connection.IsConnected();
    }

    public void Close() 
    {
        // Asynchronously disconnect from the eFTL server.
        Subscriber.connection.Disconnect();
    }

    public void Subscribe(String durable, int count)
    {
        try {
            // Create a subscription matcher for messages containing a
            // field named "type" with a value of "hello".
            //
            // When connected to an FTL channel the content matcher
            // can be used to match any field in a published message.
            // Only matching messages will be received by the
            // subscription. The content matcher can match on string
            // and integer fields, or test for the presence or absence
            // of a field by setting it's value to the boolean true or
            // false.
            //
            // When connected to an EMS channel the content matcher
            // must only contain the destination field "_dest" set to
            // the EMS topic on which to subscribe.
            //
            // To match all messages use the empty matcher "{}".
            //
            // The durable name is used to create a durable subscription.
            //
            matcher = "{\"type\":\"hello\"}";
            
            // Set the durable subscription type to shared.
            //
            options.Add(EFTL.PROPERTY_DURABLE_TYPE, DurableType.SHARED);

            // Asynchronously subscribe to matching messages.
            String subscriptionId = connection.Subscribe(matcher, durable, options, new SubscriptionListener());
            
            try {
                latch.Wait();
            } catch (Exception) {}

            // Remove the subscription.
            connection.Unsubscribe(subscriptionId);
        }
        catch (Exception e) 
        {
            Console.WriteLine(e.ToString());
        }
    }

    public static void Main (string []args) {

        for (int i = 0; i < args.Length; i++) {
            if (args[i].Equals("--username") || args[i].Equals("-u")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    username = args[++i];
                } else {
                    Usage();
                }
            } else if (args[i].Equals("--password") || args[i].Equals("-p")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    password = args[++i];
                } else {
                    Usage();
                }
            } else if (args[i].Equals("--name") || args[i].Equals("-n")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    durable = args[++i];
                } else {
                    Usage();
                }
            } else if (args[i].Equals("--count") || args[i].Equals("-c")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    count = Int32.Parse(args[++i]);
                } else {
                    Usage();
                }
            } else if (args[i].StartsWith("-")) {
                Usage();
            } else {
                url = args[i];
            }
        }

        Console.Write("#\n# Subscriber\n#\n# {0}\n#\n", EFTL.GetVersion());

        try {
            Subscriber client = new Subscriber(url, username, password);
            
            try {

                if (client.IsConnected())
                {
                    latch.Reset();
                    client.Subscribe(durable, count);            
                }
            }
            catch (Exception) {}
            finally {
                client.Close();
            }
        }
        catch (Exception e) 
        {
            Console.WriteLine(e.ToString());
        }
    }
}

